In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 8.9 MB 21.2 MB/s 
     |████████████████████████████████| 49 kB 5.1 MB/s 


$Exercise\;2$ \\
$The\;given\;linear\;programming\;problem\;is:$
$$min\; c^Tx,$$
$$s.t\; Ax ≤ b$$
$$l ≤ x ≤ u.$$



$b = [35.4, 47.9, 41.43, 22.7]$ \\
$c = [−9.6, −2.85, 41.5, 33.05, −0.75, 20.5, −0.35, −31.02, 50.65, 6.125];$ \\
$$
	A= \begin{bmatrix} 
	0 & 1 & -2 & 0 & 1 & -2 & 0 & 1 & 1 & -1 \\
	-1 & 1 & -3 & 2.5 & 0 & 1.4 & 0 & -2 & 1 & -1.5\\
	-0.45 & 4.78 & -2 & -3.22 & -1.3 & 1 & -1 & 0 & 0 & 0.25 \\
   1 & 0 & 0 & 0 & 1 & 1 & 0 & 2 & 3 & 0 \\
	\end{bmatrix}
	\quad
	$$
$l =  [−∞, −∞, 0, 0, −67, 0, −23, −5, 6, 1],$ \\
$u = [4, ∞, +∞, 2, 20, 55, 23, 15, +∞, +∞]$

$1.\;Solving\;the\;given\;problem:$

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
model = ConcreteModel()

In [ ]:
N = 10
M = 4  

In [ ]:
obj_coef_c = np.array([-9.6, -2.85,41.5,33.05, -0.75, 20.5, -0.35, -31.02, 50.65, 6.125])

In [ ]:
constr1_coef_A = np.array([[0, 1, -2, 0, 1, -2, 0, 1, 1, -1], [-1, 1, -3, 2.5, 0, 1.4, 0, -2, 1, -1.5], [-0.45, 4.78, -2, -3.22, -1.3, 1, -1, 0, 0, 0.25], [1, 0, 0, 0, 1, 1, 0, 2, 3, 0]]) 
constr_rhs_b =  np.array([35.4, 47.9, 41.43, 22.7])

In [ ]:
lower_bound = np.array([-np.inf, -np.inf, 0, 0, -67, 0, -23, -5, 6, 1])
upper_bound = np.array([4, np.inf, np.inf, 2, 20, 55, 23, 15, np.inf, np.inf])

In [ ]:
col_indices = np.arange(N)
print(col_indices)

[0 1 2 3 4 5 6 7 8 9]


In [ ]:
row_indices = np.arange(M)
print(row_indices)

[0 1 2 3]


In [ ]:
model.x = Var(col_indices)

In [ ]:
model.constraints = ConstraintList()

In [ ]:
for i in row_indices:
 model.constraints.add(sum(constr1_coef_A[i][j]*model.x[j] for j in col_indices) <= constr_rhs_b[i])

In [ ]:
for j in col_indices:
  model.x[j].setlb(lower_bound[j])
  model.x[j].setub(upper_bound[j])

In [ ]:
model.objective = Objective(expr = sum(obj_coef_c[j]*model.x[j] for j in col_indices), sense=minimize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9}

1 Var Declarations
    x : Size=10, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :  -inf :  None :   4.0 : False :  True :  Reals
          1 :  -inf :  None :   inf : False :  True :  Reals
          2 :   0.0 :  None :   inf : False :  True :  Reals
          3 :   0.0 :  None :   2.0 : False :  True :  Reals
          4 : -67.0 :  None :  20.0 : False :  True :  Reals
          5 :   0.0 :  None :  55.0 : False :  True :  Reals
          6 : -23.0 :  None :  23.0 : False :  True :  Reals
          7 :  -5.0 :  None :  15.0 : False :  True :  Reals
          8 :   6.0 :  None :   inf : False :  True :  

In [ ]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 160837 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -196.378974895398
  Upper bound: -196.378974895398
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 11
  Number of nonzeros: 29
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.011896371841430664
# ----------------------------------------------------------
#   Solution Information
# ----------

In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  -196.3789748953975

Decision Variables
x[ 0 ] =  4.0
x[ 1 ] =  5.8347280334728
x[ 2 ] =  0.0
x[ 3 ] =  0.0
x[ 4 ] =  -29.3
x[ 5 ] =  0.0
x[ 6 ] =  23.0
x[ 7 ] =  15.0
x[ 8 ] =  6.0
x[ 9 ] =  1.0

Constraints
constraints : Size=4
    Key : Lower : Body               : Upper
      1 :  None : -3.465271966527201 :  35.4
      2 :  None :  -23.6652719665272 :  47.9
      3 :  None :  41.42999999999998 : 41.43
      4 :  None :               22.7 :  22.7


$1.\; The\;optimal\;solution\;is: $ \\
$$x_1 =  4.0,\;x_2 =  5.8347280334728,\;x_3 =  0.0,\;x_4 =  0.0,\;x_5 =  -29.3,\;x_6 =  0.0,\;x_7 =  23.0,\;x_8 =  15.0,\;x_9 =  8.0,\;x_10 =  1.0,\;x_11 =  5.0,\; x_12 =  2.0$$
$The\;value\;of\;the\;objective\;function\;at\;optimal\;solution\;is:\;-196.3789748953975$

$2.\; Only\;the\;following\;two\;constraints\;are\;active\;since,\;equality\;at\;optimal\;soution\;only\;attained\;by\;these\;two\;constraints -0.45x_1 + 4.78x_2 - 2.0x_3 - 3.22x_4 - 1.3x_5 + x_6 - x_7 + 0.25x_{10} \leq 41.43;$ \
$x_1 + x_5 + x_6 + 2x_8 + 3x_9 \leq22.7;$ \\
$The\;rest\;of \;the\;constraints\;are\;inactive.$